clang/test/CodeGenCXX/debug-info-structured-binding-bitfield.cpp

```diff
@@ -8,8 +8,8 @@ struct S0 {
 // CHECK-LABEL: define dso_local void @_Z3fS0v
 // CHECK:                        alloca %struct.S0, align 4
 // CHECK-NEXT:    [[TMP0:%.*]] = alloca %struct.S0, align 4
-// CHECK:         call void @llvm.dbg.declare(metadata ptr [[TMP0]], metadata [[S0_A:![0-9]+]], metadata !DIExpression())
-// CHECK-NEXT:    call void @llvm.dbg.declare(metadata ptr [[TMP0]], metadata [[S0_B:![0-9]+]], metadata !DIExpression(DW_OP_plus_uconst, 2))
+// CHECK:         call void @llvm.dbg.declare(metadata ptr [[TMP0]], metadata [[S0_A:![0-9]+]], metadata !DIExpression(DW_OP_bit_piece, 16, 0))
```

Comment:
We generally don't use DW_OP_bit_piece directly in LLVM IR. The correct thing to do is to use DW_OP_LLVM_fragment instead, which then gets lowered to a bit_piece in AsmPrinter.

Reply:
The LLVM langref says "Note that contrary to DW_OP_bit_piece, the offset is describing the location within the described source variable". What I think this means is that while DW_OP_bit_piece means "you get the source variable by extracting these bits of this location", DW_OP_LLVM_fragment means "the value of this location is these bits of the source variable".

Trying out DW_OP_LLVM_fragment in
```
struct {
  unsigned int a : 8;
  unsigned int b : 8;
} X = { 1, 2 };

int main() {
  auto [a, b] = X;
  return a + b;
}
```
the resulting dwarf info looks pretty strange
```
 <2><7f>: Abbrev Number: 7 (DW_TAG_variable)
    <80>   DW_AT_location    : 4 byte block: 93 1 91 78         (DW_OP_piece: 1; DW_OP_fbreg: -8)
    <85>   DW_AT_name        : (indirect string, offset: 0x9c): a
    <89>   DW_AT_decl_file   : 1
    <8a>   DW_AT_decl_line   : 7
    <8b>   DW_AT_type        : <0x5f>
 <2><8f>: Abbrev Number: 7 (DW_TAG_variable)
    <90>   DW_AT_location    : 6 byte block: 93 1 91 78 93 1    (DW_OP_piece: 1; DW_OP_fbreg: -8; DW_OP_piece: 1)
    <97>   DW_AT_name        : (indirect string, offset: 0xab): b
    <9b>   DW_AT_decl_file   : 1
    <9c>   DW_AT_decl_line   : 7
    <9d>   DW_AT_type        : <0x5f>
```
lldb says (at line 8)
```
(lldb) p a
(unsigned int) 513
(lldb) p b
(unsigned int) 256
```
gdb appears to not like this dwarf info and just says
```
(gdb) p a
$1 = <optimized out>
(gdb) p b
$2 = <optimized out>
```

Reply:
https://dwarfstd.org/issues/191025.1.html
https://dwarfstd.org/issues/161206.2.html

I think you are right about the semantics of `DW_OP_bit_piece`.
I also think I misunderstood the problem the first time around.

Here a and b are not fragments of a large struct, but we want to express the opposite. They point into the same alloca. Here's incorrect ToT output:

Reply:
```
  %0 = alloca %struct.anon, align 4
  store i32 0, ptr %retval, align 4
  tail call void @llvm.dbg.declare(metadata ptr %0, metadata !23, metadata !DIExpression()), !dbg !25
  tail call void @llvm.dbg.declare(metadata ptr %0, metadata !26, metadata !DIExpression(DW_OP_plus_uconst, 1)), !dbg !27

...
!23 = !DILocalVariable(name: "a", scope: !18, file: !5, line: 2, type: !24)
!26 = !DILocalVariable(name: "b", scope: !18, file: !5, line: 2, type: !24)
```

Reply:
So we could add DW_OP_bit_piece to LLVM, or we could emit
```
!DIExpression()
!DIExpression(DW_OP_shr 8)
```
Which, I believe, would be supported by LLVM today without any modifications?

Reply:
Regardless of which variant we choose, we probably need to add explicit support for it to SROA.cpp to make sure it gets translated correctly when the alloca gets split up again.

Reply:
The SROA support doesn't need to hold up this patch of course, I'm just mentioning it.

Reply:
> So we could add DW_OP_bit_piece to LLVM, or we could emit
> 
> ```
> !DIExpression()
> !DIExpression(DW_OP_shr 8)
> ```
> 
> Which, I believe, would be supported by LLVM today without any modifications?

That would handle the bitfield offset, but wouldn't get the bitfield size right (we could use the existing CreateBindingDeclType but that can only handle cases where the bitfield size exactly matches an integer type size). 

> Regardless of which variant we choose, we probably need to add explicit support for it to SROA.cpp to make sure it gets translated correctly when the alloca gets split up again.

I'm currently working on this, should have a patch for it soon (though it looks like we only need changes in DIExpression::createFragmentExpression).